core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java

```diff
@@ -520,27 +512,17 @@ public void returnTransport(TTransport tsc) {
 
     ArrayList<CachedConnection> closeList = new ArrayList<>();
 
-    synchronized (this) {
+    Lock lock = getLock(ctsc.getCacheKey());
```

Comment:
This is at least once place that probably needs to handle concurrent shutdown.  If we try to get the locks and cache and its shutdown, then maybe jsut close the connection?

Reply:
I'm not entirely certain that I managed to address this sufficiently, but let me know what you think.

Reply:
May have handled this with the changes in lbschanno#3